src/openzaak/components/documenten/tasks.py

```diff
@@ -0,0 +1,377 @@
+        EnkelvoudigInformatieObjectCanonical.objects.bulk_create(
+            [row.instance.canonical for row in batch if row.instance is not None]
+        )
+    except DatabaseError as e:
+        for row in batch:
+            row.processed = True
+            row.comment = f"Unable to load row due to batch error: {str(e)}"
+
+        raise e
+
+    try:
+        eios = EnkelvoudigInformatieObject.objects.bulk_create(
+            [row.instance for row in batch if row.instance is not None]
+        )
+    except DatabaseError as e:
+        for row in batch:
+            row.processed = True
+            row.comment = f"Unable to load row due to batch error: {str(e)}"
+
+        raise e
+
+    # reuse created instances
+    for row in batch:
+        if row.failed:
+            continue
+
+        instance = next(
+            (
+                eio
+                for eio in eios
+                if row.instance and str(eio.uuid) == str(row.instance.uuid)
+            ),
+            None,
+        )
+
+        row.instance = instance
+
+        if not row.zaak_id:
+            row.processed = True
+            row.succeeded = True if instance and instance.pk is not None else False
+            continue
+
+        # Note that ZaakInformatieObject's will not be created using
+        # `bulk_create` (see queryset MRO).
+        zaak_eio = ZaakInformatieObject(
+            zaak_id=zaak_uuids.get(row.zaak_id),
+            informatieobject=instance.canonical,
+            aard_relatie=RelatieAarden.from_object_type("zaak"),
+        )
+
+        try:
+            zaak_eio.save()
+        except DatabaseError as e:
+            row.processed = True
+            row.comment = (
+                f"Unable to couple row {row.row_index} to ZAAK {row.zaak_id}:"
+                f"\n {str(e)}"
+            )
+
+            for _row in batch:
+                if _row.row_index == row.row_index:
+                    continue
+
+                _row.processed = True
+                _row.comment = (
+                    f"Unable to load row due to database error on row {row.row_index}"
+                )
+
+            raise e
+
+        row.processed = True
+        row.succeeded = True
+
+
+# TODO: make this more generic?
+@celery_app.task(bind=True)
+@task_locker
+def import_documents(self, import_pk: int) -> None:
+    import_instance = Import.objects.get(pk=import_pk)
+
+    file_path = import_instance.import_file.path
+
+    import_instance.total = get_total_count(file_path)
+    import_instance.started_on = timezone.now()
+    import_instance.status = ImportStatusChoices.active
+    import_instance.save(update_fields=["total", "started_on", "status"])
+
+    batch: list[DocumentRow] = []
+    batch_size = settings.IMPORT_DOCUMENTEN_BATCH_SIZE
+
+    zaak_uuids = {str(uuid): id for uuid, id in Zaak.objects.values_list("uuid", "id")}
+    eio_uuids = [
+        str(uuid) for uuid in EnkelvoudigInformatieObject.objects.values_list("uuid")
+    ]
```

Comment:
I hope it won't give us OOM error. 
Let's test it with million of zaken in the test environment when the PR is merged?

Reply:
Yeah I think we need to test this with test data that is similar to that of what will be used in production environments